specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1080 @@
+> Params of atomic execution methods
+```go
+// LockParams wraps serialized params from a message with the requested methodnum.
+type LockParams struct {
+    Method abi.MethodNum
+    Params []byte
+}
+
+// UnlockParams identifies the input params of a message
+// along with the output state to merge.
+type UnlockParams struct {
+    Params *LockParams
+    State  []byte
+}
+
+// MergeParams wraps locked state to merge in params.
+type MergeParams struct {
+    State []byte
+}
+```
+
+#### Implementation of the protocol
+The atomic execution protocol consists of the following phases, which, combined, resemble a two-phase commit protocol with the `SCA` of the least common ancestor/parent serving as a coordinator:
+
+- __Initialization__: To signal the start of an atomic execution, the parties involved interact off-chain to agree on the specific execution they want to perform and the input state it will involve. 
+    - To start the execution, each party needs to lock, in their subnet, the state that will be used as input for the execution. This is done by sending a message to the `Lock()` function to the address of the `LockableActor` involved in the execution indicating in the `LockParams` the `MethodNum` and `Params` the parties agreed on for the atomic execution. `Lock()` returns the `CID` of the locked state and persists it in the corresponding `LockedStateRegistry` of the actor. Locking the state prevents new messages from affecting the state and leading to inconsistencies when the output state is migrated back. From now on, the actor won't accept any message involving the state locked. The locking of the input state in each subnet signals the beginning of the atomic execution.
+    - One of the parties needs to initialize the execution by sending an `InitAtomicExec` message to the `SCA` of the parent responsible for orchestrating the execution specifying the `AtomicExecParams` (i.e. the list of messages and a map with the HA address and the `CID` of the input state for all the parties involved). If the initiator already has performed the off-chain execution (see next bullet), it can also submit the CID of its output state as part of the initialization message.
+    - In its current implementation `SCA` verifies that is the common parent of all parties to accept the execution. To uniquely identify it computes the `CID` of the atomic execution (by computing the `CID` the `AtomicExecParams`). This `CID` identifies the atomic execution in the `SCA` through all its life. If the `InitAtomicExec` message succeed a new atomic execution is spawned for that CID in an `ExecInitialized` state.
+> Note: At this point there's no deduplication of atomic executions in the same `SCA`. An atomic execution with the same `AtomicExecParams` can't be run twice in the same `SCA`. This is a limitation of the MVP implementation and it will be fixed in the next iteration by adding a nonce (see [issue](https://github.com/filecoin-project/eudico/issues/147).
```

Comment:
```suggestion
> Note: At this point there's no deduplication of atomic executions in the same `SCA`. An atomic execution with the same `AtomicExecParams` can't be run twice in the same `SCA`. This is a limitation of the MVP implementation and it will be fixed in the next iteration by adding a nonce (see [issue](https://github.com/filecoin-project/eudico/issues/147)).
```